File: src/commonMain/kotlin/com/xebia/functional/embeddings/OpenAIEmbeddings.kt
package com.xebia.functional.embeddings

import arrow.fx.coroutines.parMap
import arrow.resilience.retry
import com.xebia.functional.env.OpenAIConfig
import com.xebia.functional.llm.openai.EmbeddingRequest
import com.xebia.functional.llm.openai.OpenAIClient
import com.xebia.functional.llm.openai.RequestConfig
import io.github.oshai.KLogger
import kotlin.time.ExperimentalTime

@ExperimentalTime
class OpenAIEmbeddings(
  private val config: OpenAIConfig,
  private val oaiClient: OpenAIClient,
  private val logger: KLogger
) : Embeddings {

  override suspend fun embedDocuments(
    texts: List<String>,
    chunkSize: Int?,
    requestConfig: RequestConfig
  ): List<Embedding> =
    chunkedEmbedDocuments(texts, chunkSize ?: config.chunkSize, requestConfig)

  override suspend fun embedQuery(text: String, requestConfig: RequestConfig): List<Embedding> =
    if (text.isNotEmpty()) embedDocuments(listOf(text), null, requestConfig) else emptyList()

  private suspend fun chunkedEmbedDocuments(
    texts: List<String>,
    chunkSize: Int,
    requestConfig: RequestConfig
  ): List<Embedding> =
    if (texts.isEmpty()) emptyList()
    else texts.chunked(chunkSize)
      .parMap { createEmbeddingWithRetry(it, requestConfig) }
      .flatten()

  private suspend fun createEmbeddingWithRetry(texts: List<String>, requestConfig: RequestConfig): List<Embedding> =
    kotlin.runCatching {
      config.retryConfig.schedule()
        .log { retriesSoFar, _ -> logger.warn { "Open AI call failed. So far we have retried $retriesSoFar times." } }
        .retry {
          oaiClient.createEmbeddings(EmbeddingRequest(requestConfig.model.name, texts, requestConfig.user.id))
            .data.map { Embedding(it.embedding) }
        }
    }.getOrElse {
      logger.warn { "Open AI call failed. Giving up after ${config.retryConfig.maxRetries} retries" }
      throw it
    }
}

File: scala/src/main/scala/com/xebia/functional/embeddings/openai/OpenAIEmbeddings.scala
package com.xebia.functional.scala.embeddings.openai

import scala.concurrent.duration.FiniteDuration

import cats.*
import cats.data.NonEmptyChain
import cats.effect.*
import cats.syntax.all.*

import com.xebia.functional.scala.config.OpenAIConfig
import com.xebia.functional.scala.embeddings.Embeddings
import com.xebia.functional.scala.embeddings.models.Embedding
import com.xebia.functional.scala.embeddings.openai.models.RequestConfig
import com.xebia.functional.scala.llm.openai.OpenAIClient
import com.xebia.functional.scala.llm.openai.models.EmbeddingRequest
import com.xebia.functional.scala.llm.openai.models.EmbeddingResult
import org.typelevel.log4cats.Logger
import retry.RetryDetails.*
import retry.RetryPolicies.*
import retry.*

class OpenAIEmbeddings[F[_]: MonadThrow: Sleep: Parallel](config: OpenAIConfig, oaiClient: OpenAIClient[F], logger: Logger[F]) extends Embeddings[F]:

  def embedQuery(text: String, rc: RequestConfig): F[Vector[Embedding]] =
    if (text.nonEmpty) embedDocuments(List(text), None, rc)
    else Monad[F].pure(Vector.empty[Embedding])

  def embedDocuments(texts: List[String], chunkSize: Option[Int], rc: RequestConfig): F[Vector[Embedding]] =
    chunkedEmbedDocuments(texts, chunkSize.getOrElse(config.chunkSize), rc)

  private def chunkedEmbedDocuments(
      texts: List[String],
      chunkSize: Int,
      rc: RequestConfig
  ): F[Vector[Embedding]] =
    if (texts.isEmpty) Monad[F].pure(Vector.empty)
    else {
      val batches = texts.sliding(chunkSize, chunkSize)
      batches.toList
        .parTraverse(batch => embedWithRetry(batch, rc))
        .map(vectors => vectors.reduce(_ ++ _))
    }

  private def embedWithRetry(texts: List[String], rc: RequestConfig): F[Vector[Embedding]] =
    retryingOnAllErrors[EmbeddingResult](
      policy = limitRetries[F](config.maxRetries) join exponentialBackoff[F](config.backoff),
      onError = logError
    )(oaiClient.createEmbeddings(EmbeddingRequest(rc.model.name, texts, rc.user.asString)))
      .map(_.data.map(result => Embedding(result.embedding)))

  private def logError(err: Throwable, details: RetryDetails): F[Unit] = details match {
    case WillDelayAndRetry(nextDelay: FiniteDuration, retriesSoFar: Int, cumulativeDelay: FiniteDuration) =>
      logger.warn(s"Open AI call failed. So far we have retried $retriesSoFar times.")
    case GivingUp(totalRetries: Int, totalDelay: FiniteDuration) =>
      logger.warn(s"Open AI call failed. Giving up after $totalRetries retries")
  }
